File: RegionTypes/EqualsRegionType.py
from typing import List

from z3 import *

from RegionTypes.RegionType import RegionType


class EqualsRegionType(RegionType):
    """
    The EqualsRegionType is for when the value is exactly equal to a certain number.
    So 2/4 is exactly 2 or exactly 4
    And 3/5/7 is exactly 3, 5, or 7
    Note that =2 and =4 also fits in here, just that it's only = to one number.
    """
    def __init__(self, vals: List):
        self.vals = [int(i) for i in vals]

    def get_expr(self, var: Int):
        return Or([var == val for val in self.vals])

File: RegionTypes/EqualsRegionType.cpp
//
// Created by cppac on 4/8/2023.
//

#include "EqualsRegionType.h"

// I didn't like having this in the header file even if it's an empty code block.
EqualsRegionType::EqualsRegionType(int validCount) : EqualsRegionType(&validCount, 1) {

}

EqualsRegionType::EqualsRegionType(const int *validCounts, size_t arrSize) : countSize(arrSize){
    if(countSize == 0){
        throw std::invalid_argument("EqualsRegionType needs to be equal to at least one number.");
    }
    this->validCounts = new int[arrSize];
    for(size_t i = 0; i < countSize; i++){
        this->validCounts[i] = validCounts[i];
    }
}

EqualsRegionType::~EqualsRegionType() {
    delete[] validCounts;
}

z3::expr EqualsRegionType::getExpr(const z3::expr& var) {
    // Constructor guarantees that countSize will always be nonzero
    z3::expr out = (var == this->validCounts[0]);
    for(size_t i = 1; i < countSize; i++){
        out = out || (var == this->validCounts[i]);
    }
    return out;
}
